### PR TITLE
Rust: New query for bad 'ctor' initialization

### PR DESCRIPTION
**Pull request by geoffw0**
New (requested) query for bad [`ctor`](https://github.com/mmastrac/rust-ctor?tab=readme-ov-file#rust-ctor) initialization.  See the `.qhelp` for an explanation of what this is.  Note three limitations:
- results are perhaps more likely to be more about correctness / reliability than security (in my opinion).  There is a suitable CWE though.
- no transitive results through calls (yet), because we don't have call target resolution (yet, as far as I'm aware).
- limited testing, because we don't yet have any real-world databases that use `#[ctor]` / `#[dtor]`.

TODO:
- [x] code review
- [x] docs review

### CODE REVIEW

**Comment by paldepind** [comment]
> results are perhaps more likely to be more about correctness / reliability than security (in my opinion). There is a suitable CWE though.

Sounds right to me.

> no transitive results through calls (yet), because we don't have call target resolution (yet, as far as I'm aware).

For global data flow we [use the resolved paths](https://github.com/github/codeql/blob/main/rust/ql/lib/codeql/rust/dataflow/internal/DataFlowImpl.qll#L551) that the extractor gives. But I don't think we have a suitable API for using that in queries yet. In C we have `getTarget` on function calls. Is that the thing that we'd want for queries?

**Review by paldepind — DISMISSED** [review]
Looks good to me 😄

**Comment by geoffw0** [comment]
> In C we have getTarget on function calls. Is that the thing that we'd want for queries?

Yep - or `getStaticTarget()` in Swift.  @andersfugmann pointed out that even a placeholder predicate would allow us to fill the gaps in this query implementation.

**Review by mchammer01 — DISMISSED** [review]
LGTM 💖 
Left some comments and suggestions for your consideration (but these are minor, and I didn't want to block this PR).

File: rust/ql/src/queries/security/CWE-696/BadCtorInitialization.qhelp [hunk 1]
```diff
@@ -0,0 +1,44 @@
+<!DOCTYPE qhelp PUBLIC
+  "-//Semmle//qhelp//EN"
+  "qhelp.dtd">
+<qhelp>
+<overview>
+
+<p>
+Calling functions and methods in the Rust <code>std</code> library from a <code>#[ctor]</code> or <code>#[dtor]</code> function is not safe. This is because the <code>std</code> library only guarantees stability and portability between the beginning and end of <code>main</code>, whereas <code>#[ctor]</code> functions are called before <code>main</code>, and <code>#[dtor]</code> functions are called after it.
```
[the remark this answers is not in the document]
Comment (geoffw0) [reply rust/ql/src/queries/security/CWE-696/BadCtorInitialization.qhelp:8]: I'm pretty neutral on this one.  Accepted.

**Comment by geoffw0** [comment]
Thanks for the docs review, all changes accepted.

**Review by mchammer01 — DISMISSED** [review]
Thanks for looking at my comments/suggestions @geoffw0 🙇🏻 
LGTM :shipit:

**Review by paldepind — APPROVED** [review]

**Pull request merged**